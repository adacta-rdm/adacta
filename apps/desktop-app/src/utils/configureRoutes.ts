import type { RouteConfig, RouteObject } from "found";
import { Redirect } from "found";

type Route = string; //(typeof routes)[number];

export function configureRoutes(input: Record<Route, RouteObject | string>): RouteConfig {
	// Similar to the structure of the input, but with the routes with their correct relative paths and children set.
	// This way, we can conveniently query the object to find the parent of a route.
	// The root RouteObject is the one with the path "/" and we'll return the children property of that object.
	const root = { route: "/", path: "/", children: [] as IRouteObject[] };
	const routeNodes: Record<string, IRouteObject> = { "/": root };

	// Need to sort the input, so that the parents are processed first
	const sortedRoutes = Object.entries(input).sort(([a], [b]) => {
		// Split each route into segments based on the '/' delimiter
		const aSegments = a.split("/");
		const bSegments = b.split("/");

		// Iterate over each segment in the routes
		// The loop runs for the length of the shorter route to prevent out-of-bounds access
		const l = Math.min(aSegments.length, bSegments.length);
		for (let i = 0; i < l; i++) {
			// Check if the current segment in each route is a variable (begins with ':')
			const aIsVariable = aSegments[i][0] === ":";
			const bIsVariable = bSegments[i][0] === ":";

			// If one route's segment is variable and the other's is static, the static one should come first
			if (aIsVariable && !bIsVariable) {
				return 1; // a comes after b
			} else if (!aIsVariable && bIsVariable) {
				return -1; // a comes before b
			} else {
				// If both segments are either static or variable, compare them alphabetically
				const comparison = aSegments[i].localeCompare(bSegments[i]);
				// If the segments are not equal, return the comparison result
				if (comparison !== 0) {
					return comparison;
				}
			}
		}

		// If all segments are equal, compare the routes by length
		// This ensures that shorter routes come before longer ones
		return aSegments.length - bSegments.length;
	});
	for (const [route, cfg] of sortedRoutes) {
		if (typeof cfg === "string") {
			const redirect = new Redirect({ from: route, to: cfg });
			root.children.push(redirect as any);
			continue;
		}

		// Determine the parent belonging to this route
		// We initialize it to the root, so that if we can't find a parent, we'll use the root
		let parent: IRouteObject = root;
		const parts = route.split("/");
		// Normalize the parts, so that we don't have empty strings
		// .filter((part) => part !== "");

		while (parts.length > 0) {
			parts.pop();
			const parentPath = `${parts.join("/")}`;
			const parentCandidate = routeNodes[parentPath];

			// If the parent is a string, it's a redirect, and we can't use it as a parent
			if (typeof parentCandidate === "object") {
				parent = parentCandidate;
				break;
			}
		}

		// console.log(route, parent);

		// Determine the relative path of the route based on its parent.
		let path = `${route.slice(parent.route.length)}`;
		if (path[0] !== "/") path = `/${path}`;

		const { getData, Component } = cfg;
		const node: IRouteObject = {
			path,
			route,
			getData,
			Component,
		};

		if (!parent.children) parent.children = [];
		parent.children.push(node);
		routeNodes[route] = node;
	}

	return root.children as RouteConfig;
}

interface IRouteObject {
	/**
	 * The absolute path to the route.
	 */
	route: string;
	path: string;

	getData?: RouteObject["getData"];
	Component?: RouteObject["Component"];

	// Component?: (...args: unknown[]) => unknown;
	children?: IRouteObject[];
}
