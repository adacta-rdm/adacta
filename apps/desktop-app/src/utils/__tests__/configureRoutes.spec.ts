import type { RouteObject } from "found";
import { describe, test, expect } from "vitest";

import { configureRoutes } from "../configureRoutes";

describe("configureRoutes()", () => {
	test("snapshot", () => {
		const input: Record<string, RouteObject | string> = {
			"/": "/repositories", // Redirect
			"/login": {},
			"/repositories": {},
			"/repositories/": {}, // Index route
			"/repositories/new": {}, // New repository
			"/long/path/to/somewhere": {}, // Route with intermediates not explicitly defined
		};

		const output = configureRoutes(input);

		expect(output).toMatchSnapshot();
	});

	test("static path names have precedence over variable path parts", () => {
		const input: Record<string, RouteObject | string> = {
			"/": "/repositories", // Redirect
			"/login": {},
			"/user/:id": {}, // Variable path
			"/user/static": {}, // Static path
		};

		const output = configureRoutes(input);

		// Find the indices of the static and variable paths
		const staticPathIndex = output.findIndex((route) => route.path === "/user/static");
		const variablePathIndex = output.findIndex((route) => route.path === "/user/:id");

		// The static path should come before the variable path
		expect(staticPathIndex).toBeLessThan(variablePathIndex);
	});
});
