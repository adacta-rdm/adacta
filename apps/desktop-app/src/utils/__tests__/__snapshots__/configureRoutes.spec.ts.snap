// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`configureRoutes() > snapshot 1`] = `
[
  Redirect {
    "path": "/",
    "status": undefined,
    "to": "/repositories",
  },
  {
    "Component": undefined,
    "getData": undefined,
    "path": "/login",
    "route": "/login",
  },
  {
    "Component": undefined,
    "getData": undefined,
    "path": "/long/path/to/somewhere",
    "route": "/long/path/to/somewhere",
  },
  {
    "Component": undefined,
    "children": [
      {
        "Component": undefined,
        "getData": undefined,
        "path": "/",
        "route": "/repositories/",
      },
      {
        "Component": undefined,
        "getData": undefined,
        "path": "/new",
        "route": "/repositories/new",
      },
    ],
    "getData": undefined,
    "path": "/repositories",
    "route": "/repositories",
  },
]
`;
